aii-pxelinux/src/main/perl/pxelinux.pm

```diff
@@ -397,93 +502,115 @@ sub pxelink
         $path = $this_app->option (LOCALBOOT);
         unless ($path =~ m{^([-.\w]+)$}) {
             $this_app->error ("Unexpected BOOT configuration file");
-            return -1;
+            return 0;
         }
         $path = $1;
         $this_app->debug (5, "Local booting from $path");
     } elsif ($cmd eq RESCUE || $cmd eq LIVECD || $cmd eq FIRMWARE) {
-        $path = link_filepath($cfg, $cmd);
-        if (! -f $path) {
+        $path = link_filepath($cfg, $cmd, $variant);
+        # Testing file existence in unit tests may fail... and file_exists()
+        # ignores NoAction flag as of 16.12
+        if (! CAF::Path->file_exists($path) && ! $CAF::Object::NoAction ) {
             my $fqdn = get_fqdn($cfg);
             $this_app->error("Missing $cmd config file for $fqdn: $path");
-            return -1;
+            return 0;
         }
         $this_app->debug (5, "Using $cmd from: $path");
     } elsif ($cmd eq INSTALL) {
-        $path = filepath ($cfg);
+        $path = filepath ($cfg, $variant);
         $this_app->debug (5, "Installing on $path");
     } else {
         $this_app->debug (5, "Unknown command");
-        return -1;
+        return 0;
     }
     # Set the same settings for every network interface that has a
     # defined IP address.
-    foreach my $st (values (%$t)) {
+    foreach my $st (values (%$interfaces)) {
         next unless $st->{ip};
-        my $dir = $this_app->option (NBPDIR);
-        my $lnname = "$dir/".hexip ($st->{ip});
+        my $dir = variant_option('nbpdir_opt', $variant);
+        my $lnname = "$dir/".hexip_filename ($st->{ip}, $variant);
         if ($cmd || ! -l $lnname) {
             if ($CAF::Object::NoAction) {
                 $this_app->info ("Would symlink $path to $lnname");
             } else {
                 unlink ($lnname);
-                # This must be stripped to work with chroot'edg
-                # environments.
+                # This must be stripped to work with chroot'ed environments.
                 $path =~ s{$dir/?}{};
                 symlink ($path, $lnname);
             }
         }
     }
-    return 0;
+
+    return 1;
+}
+
+
+# Wrapper function to call ksuserhooks() from aii-ks module.
+# The only role of this function is to ensure that ksuserhooks()
+# is always called the same way (in particular for NoAction
+# handling). Be sure to use it!
+sub exec_userhooks {
+    my ($cfg, $hook_path) = @_;
+
+    ksuserhooks ($cfg, $hook_path) unless $CAF::Object::NoAction;
```

Comment:
In fact this is what you implemented! It took me some time to realize that it worked because the method was mocked in the private version of ks.pm used for testing....